packages/plexus/src/LayoutManager/Coordinator.js

```diff
@@ -0,0 +1,346 @@
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import type { LayoutUpdate, WorkerMessage } from './types';
+import type { Edge, LayoutGraph, LayoutVertex, SizeVertex } from '../types/layout';
+import type { ErrorEvent } from '../types/ErrorEvent';
+
+import * as convCoord from './dot/conv-coord';
+import workerFactory from './layout.worker';
+import { matchEdges, matchVertices } from './match-inputs';
+
+type CurrentLayout = {
+  id: number,
+  input: {
+    edges: Edge[],
+    vertices: SizeVertex[],
+  },
+  cleaned: {
+    edges: Edge[],
+    vertices: SizeVertex[],
+  },
+  result: {
+    // TODO(joe): split edge calculations
+    // edges: LayoutEdge[],
+    graph?: LayoutGraph,
+    vertices: LayoutVertex[],
+  },
+  status: {
+    edgesWorkers?: ?Set<number>,
+    phase: 'not-started' | 'positions' | 'edges' | 'done',
+    positionsWorker?: ?number,
+  },
+};
+
+type LayoutWorker = Worker & { id: number };
+
+function cleanInput(srcEdges: Edge[], srcVertices: SizeVertex[]) {
+  // edges: Edge[], vertices: SizeVertex[]
+  const edges = srcEdges.map(({ from, to, isBidirectional }) => ({ from, to, isBidirectional }));
+  // const edges = srcEdges.map(edge => {
+  //   //({ from, to, bidirectional })
+  //   const { from, to } = edge;
+  //   const rv = { from, to };
+  //   if (edge.bidirectional != null) {
+  //     rv.bidirectional = edge.bidirectional;
+  //   }
+  //   return rv;
+  // });
+  const vertices = srcVertices.map(({ vertex: { key }, ...rest }) => ({ vertex: { key }, ...rest }));
+  return { edges, vertices };
+}
+
+function killWorker(worker: LayoutWorker) {
+  const w = worker;
+  // to make flow happy
+  const noop = () => {};
+  w.onmessage = noop;
+  // $FlowFixMe - https://github.com/facebook/flow/issues/6191
+  w.onmessageerror = noop;
+  w.onerror = noop;
+  w.terminate();
+}
+
+function findAndRemoveWorker(lists: LayoutWorker[][], worker: LayoutWorker) {
+  for (let i = 0; i < lists.length; i++) {
+    const list = lists[i];
+    const wi = list.indexOf(worker);
+    if (wi >= 0) {
+      list.splice(wi, 1);
+      return { ok: true, list };
+    }
+  }
+  return { ok: false };
+}
+
+export default class Coordinator {
+  currentLayout: ?CurrentLayout;
+  nextWorkerId: number;
+  isDisposed: boolean;
+  idleWorkers: LayoutWorker[];
+  busyWorkers: LayoutWorker[];
+  callback: LayoutUpdate => void;
+
+  constructor(callback: LayoutUpdate => void) {
+    this.callback = callback;
+    this.currentLayout = null;
+    this.nextWorkerId = 1;
+    this.isDisposed = false;
+    this.idleWorkers = [this._initWorker()];
```

Comment:
Why initialize with a worker? It looks like you just create one as they're required. Seems like introducing unnecessary state.

Reply:
It takes some time to initialize the worker. A [quick test](https://fiddle.jshell.net/1ukjdxpb/show/light/) showed around 7 - 10ms. (Most of it is in generating the blob.) Kind of point-less to not do it lazily, though.

Will change.

As a side note: the rationale was driven by the fact that the workers are pooled.

Reply:
This looks better to me 👍